File: ruby/parent-child.rb
def parent_child data
  roots = []

  tmp_hash = []

  data.each_with_index do |d, i|

    tmp_hash[d[:id]] = {
      :id => d[:id],
      :children=> []
    }

    node = tmp_hash[d[:id]]
    if d[:parent].to_i === 0
      roots.push node
    else
      parent = tmp_hash[d[:parent]]

      # puts parent
      parent[:children].push node
    end
    puts i.to_s + " roots: "
    puts roots
    # i = i + 1
  end
  # puts "!!!!!!!!!!!!"

  return roots
end

File: javascript/parent-child.js
function parent_child(tree){
  var tmp = {};
  var roots = [];

  for (var i in tree) {
    tmp[tree[i].id] = {
      'id': tree[i].id,
      'children': []
    };
  }

  // console.log(tmp);

  tree.forEach(function(current){
    var node = tmp[current.id];

    if ( parseInt(current.parent) === 0 ){
      roots.push(node);
    }else{
      var parent = tmp[current.parent];
      parent.children.push(node);
    }
  });

  return roots;
};

module.exports = parent_child;
